### PR TITLE
Add update cluster interface to the TKE driver

### PR DESCRIPTION
**Pull request by guangbochen**
**Problem:**
Support edit TKE cluster and allowing add more worker node.

**Solution:**
Added update cluster interface to the TKE driver, and the configurable parameters are listed as below:

| Variable  | Type   | Is_required  | Default  | Description  |
|---|---|---|---|---|
| secretId  | string | yes | N/A |  the secret id |
| secretKey  | string | yes | N/A |  the secret key |
| driverName  | string | yes | tencentkubernetesengine |  the secret key |
| nodeCount  |  int | yes | N/A | The number of total nodes purchased. |
| instanceType  |  string | no | S2.MEDIUM4  |  CVM Instance type |
| storageSize | int | no | 50  | Data disk size (GB), the step size is 10 |
| rootSize| int | no | 50 | System disk size. Linux system adjustment range is 20 - 50G, step size is 1  |
| clusterName |  string  | no  | none   |  cluster name  |
| clusterDesc  |  string | no | none  |  cluster description |
| projectId  | int  | no | 0  | modify the projectId that the k8s cluster belongs to in tencentCloud. |
| bandwidthType | string | no | PayByTraffic  | PayByTraffic or PayByHour  |
| bandwidth|  int  | no | 1 | Public network bandwidth (Mbps), when the traffic is charged for the public network bandwidth peak  |
| wanIp  | int | no | 1 |  the cluster master occupies the IP of a VPC subnet |
| isVpcGateway | int | no | 0 |  Whether it is a public network gateway, network gateway only in public with a public IP, and in order to work properly when under private network |

**Issues:**
https://github.com/rancher/rancher/issues/17175

### CODE REVIEW

File: driver/tke_driver.go [hunk 1]
```diff
@@ -296,36 +296,68 @@ func (d *Driver) GetDriverUpdateOptions(ctx context.Context) (*types.DriverFlags
...
+		Type:     types.StringType,
+		Password: true,
+		Usage:    "The secretKey of the cluster",
+	}
 	driverFlag.Options["cluster-id"] = &types.Flag{
 		Type:  types.StringType,
 		Usage: "The ID of the existing cluster",
 	}
-	driverFlag.Options["cluster-name"] = &types.Flag{
+	driverFlag.Options["goods-num"] = &types.Flag{
```
Comment (orangedeng) [inline driver/tke_driver.go]: I don't like the variable name `goods-num`. Can we rename it to `node-count` or something meaningful?

File: driver/tke_driver.go [hunk 2]
```diff
@@ -564,8 +603,83 @@ func getState(info *types.ClusterInfo) (*state, error) {
...
+	}
+
+	svc, err := getTKEServiceClient(state, "POST")
+	if err != nil {
+		return nil, err
+	}
+
+	logrus.Debugf("Updating config, clusterName: %s, clusterVersion: %s, new node: %v", state.ClusterName, state.ClusterVersion, newState.GoodsNum)
+
+	if newState.GoodsNum != 0 {
```
Comment (orangedeng) [inline driver/tke_driver.go]: should we compare new node count with the current node count also? I don't think we need to send update request if the node count is not changed.

File: driver/tke_driver.go [hunk 3]
```diff
@@ -851,3 +964,97 @@ func (d *Driver) GetK8SCapabilities(ctx context.Context, opts *types.DriverOptio
 	}
 	return capabilities, nil
 }
+
+func getWrapAddClusterInstancesRequest(state, newState *state) (*ccs.AddClusterInstancesRequest, error) {
```
Comment (orangedeng) [inline driver/tke_driver.go]: Should we call it `getWrapUpdateClusterInstancesRequest`? it is used when updating cluster.

File: driver/tke_driver.go [hunk 4]
```diff
@@ -296,36 +297,68 @@ func (d *Driver) GetDriverUpdateOptions(ctx context.Context) (*types.DriverFlags
...
 	}
-	driverFlag.Options["secret-id"] = &types.Flag{
+	driverFlag.Options["project-id"] = &types.Flag{
+		Type:  types.IntType,
+		Usage: "The ID of your project to use when creating a cluster",
+	}
+	driverFlag.Options["bandwidth-type"] = &types.Flag{
 		Type:  types.StringType,
-		Usage: "The secretID of the cluster",
+		Usage: "Type of bandwidth, PayByTraffic or PayByHour",
```
Comment (orangedeng) [inline driver/tke_driver.go:353]: Please add default value message to `Usage` in order to tell users which one is using if blank. Or you can just add default value for this flag.

**Review by orangedeng — COMMENTED** [review]
The only thing bugging me is all the default values of the flags. The user should know which one is using if you set the it blank. Or you can just add default value to those flags.

**Comment by loganhz** [comment]
Why do you use `nodeCount` instead of `goodsNum`?

**Comment by guangbochen** [comment]
`nodeCount` will be using for both create and update.

For editing cluster, only the below parameters are required, optional values will use stored state value from creation if is not specified.
e.g,
```
{
    "nodeCount":2,
    "secretKey":"xxxxxxx",
    "secretId":"xxxxxx",
    "driverName":"tencentkubernetesengine"
}
```
@loganhz  @orangedeng

File: driver/tke_driver.go [hunk 5]
```diff
@@ -173,16 +174,19 @@ func (d *Driver) GetDriverCreateOptions(ctx context.Context) (*types.DriverFlags
 		Usage: "The IP address range of the container pods, must not conflict with VPC CIDR",
 	}
 	driverFlag.Options["ignore-cluster-cidr-conflict"] = &types.Flag{
-		Type:  types.BoolType,
-		Usage: "Whether to ignore the ClusterCIDR conflict error, the default is 0",
+		Type:    types.BoolType,
+		Usage:   "Whether to ignore the ClusterCIDR conflict error, the default is 0",
+		Default: &types.Default{DefaultInt: 0},
```
Comment (orangedeng) [inline driver/tke_driver.go:179]: You are setting a bool type default value as int 0. It is a little bit confusing.

File: driver/tke_driver.go [hunk 6]
```diff
@@ -196,9 +200,9 @@ func (d *Driver) GetDriverCreateOptions(ctx context.Context) (*types.DriverFlags
 		Type:  types.StringType,
 		Usage: "The zone id of the cluster",
 	}
-	driverFlag.Options["goods-num"] = &types.Flag{
+	driverFlag.Options["node-count"] = &types.Flag{
 		Type:  types.IntType,
-		Usage: "The number of nodes purchased, up to 100",
+		Usage: "The number of nodes to purchase, up to 100",
```
Comment (orangedeng) [inline driver/tke_driver.go]: Please update the usage for `node-count` flag. It should be like `The node count of the cluster.`

File: driver/tke_driver.go [hunk 7]
```diff
@@ -209,32 +213,38 @@ func (d *Driver) GetDriverCreateOptions(ctx context.Context) (*types.DriverFlags
...
+		Type:    types.StringType,
+		Usage:   "See CVM Instance Configuration for details . Default: S2.MEDIUM4",
+		Default: &types.Default{DefaultString: "S2.MEDIUM4"},
 	}
 	driverFlag.Options["cvm-type"] = &types.Flag{
-		Type:  types.StringType,
-		Usage: "The type of node, the default is charged by volume ",
+		Type:    types.StringType,
+		Usage:   "The type of node, the default is charged by volume ",
+		Default: &types.Default{DefaultString: "PayByHour"},
```
Comment (orangedeng) [inline driver/tke_driver.go:228]: I don't understand the default value and the usage of this flag.

**Comment by guangbochen** [comment]
@orangedeng updated, thanks.

**Comment by loganhz** [comment]
Tested. 
I can create, scale up, scale down and remove TKE cluster.

**Comment by orangedeng** [comment]
LGTM

**Pull request merged**